docs/RFCS/20220426_rbac_extensions_system_privileges.md

```diff
@@ -0,0 +1,137 @@
+VIEWACTIVITYREDACTED / NOVIEWACTIVITYREDACTED | Replace with system level grant
+SQLLOGIN / NOSQLLOGIN | Unchanged
+VIEWCLUSTERSETTING / NOVIEWCLUSTERSETTING | Replace with system level grant 
+
+##  New System Table
+
+We will create a new “system.privileges” table.
+ `
+CREATE TABLE system.privileges (
+username STRING NOT NULL,
+object INT NOT NULL,
+object_type system.object_type NOT NULL,
+privileges VARBIT NOT NULL,
+role_options VARBIT NOT NULL,
+CONSTRAINT "primary" PRIMARY KEY (username, object, object_type)
+);`
+
+`system.object_type` will be an enum that defines the various object types that can support privileges in CockroachDB. This tentatively looks like 
+```
+CREATE TYPE system.object_type AS ENUM('system', 'jobs', 'changefeed_sink')
+```
+
+Note that this will be the first example of a system enum.
+
+
+The system.privileges table will store privileges for non-descriptor backed objects.
+We will require object_type + id to uniquely identify them. The privileges will be represented as a bit mask. We can serialize and deserialize synthetic privilege descriptors using rows in this table to leverage our existing code that manipulates PrivilegeDescriptors for grants. We can also consider making the privileges and role_options columns have type `[]STRING`, the downside here is that it takes slightly more space (probably not a concern) and that we have to add some logic to serialize the string array into a bit array which is how we currently manipulate privileges. 
+
+### Alternate considerations
+The system level grants could be stored on a PrivilegeDescriptor on the system database, however this would require us to support mutations on the system database descriptor which we currently do not do and this allows us to not perform a round trip whenever we look for a system table.
+
+We could create descriptors to represent every object we want to grant on. This is not an attractive option at first glance as these descriptors will likely only be used to store privileges and nothing more.
+
+## Syntax for System Level Privileges
+The current proposal is to support `GRANT SYSTEM [privileges...] TO [users...]`.
+We cannot do `GRANT [privileges…] ON SYSTEM TO [users…]` due to the fact that tables can be called system, this would be ambiguous. Using two words, such as SYSTEM ADMIN (can’t think of a good phrase here) would work.
+
+Proposed new system level privileges:
+VIEWACTIVITY (this can be broken down into more privileges)
+VIEWACTIVITYREDACTED
+CANCELQUERY
+MODIFYCLUSTERSETTING
+VIEWCLSUTERSETTING
+VIEWSYSTEMTABLES (can be broken down further if necessary)
+
+
+The syntax is up in the air however, the grant would create or modify the entry in the system.privileges table.
+
+Example: 
+`GRANT SYSTEM MODIFYCLUSTERSETTING TO foo` 
+
+Results in the row
+(foo, 1, “system”, 16384, 0)
+
+The bitmask value for MODIFYCLUSTERSETTING is assumed to be 2 << 13 = 16384 in the above example.
```

Comment:
nit: do bitmasks usually make more sense as `1 << 14` and `1 << 15`? that way you're moving 0b1 which is easier to visualise than 0b10, unless i'm missing something.

Reply:
Done